File: iosApp/iosApp/Models/ApiModels.swift
import Foundation

// MARK: - Auth

struct AuthResponse {
    let user: ApiUser
    let accessToken: String
    let refreshToken: String
    let isNewUser: Bool

    static func from(_ dict: [String: Any]) throws -> AuthResponse {
        guard let userDict = dict["user"] as? [String: Any],
              let at = dict["access_token"] as? String,
              let rt = dict["refresh_token"] as? String else {
            throw ApiError.decode("Invalid auth response")
        }
        return AuthResponse(
            user: ApiUser.from(userDict),
            accessToken: at,
            refreshToken: rt,
            isNewUser: dict["is_new_user"] as? Bool ?? false
        )
    }
}

struct TokenPair {
    let accessToken: String
    let refreshToken: String

    static func from(_ dict: [String: Any]) throws -> TokenPair {
        guard let at = dict["access_token"] as? String,
              let rt = dict["refresh_token"] as? String else {
            throw ApiError.decode("Invalid token pair")
        }
        return TokenPair(accessToken: at, refreshToken: rt)
    }
}

struct ApiUser {
    let id: String
    let name: String?
    let email: String?
    let phone: String?
    let avatarUrl: String?
    let locale: String?
    let addresses: [ApiAddress]
    let accounts: [ApiAccount]

    var displayName: String? { name }

    var primaryAddress: ApiAddress? { addresses.first { $0.isPrimary } ?? addresses.first }

    var initials: String {
        let parts = (name ?? "").split(separator: " ")
        if parts.count >= 2 {
            return "\(parts[0].prefix(1))\(parts[1].prefix(1))".uppercased()
        }
        return String((name ?? "?").prefix(2)).uppercased()
    }

    static func from(_ dict: [String: Any]) -> ApiUser {
        let addrs = (dict["addresses"] as? [[String: Any]] ?? []).compactMap { ApiAddress.from($0) }
        let accs = (dict["accounts"] as? [[String: Any]] ?? []).compactMap { ApiAccount.from($0) }
        return ApiUser(
            id: dict["id"] as? String ?? "",
            name: dict["name"] as? String ?? dict["display_name"] as? String,
            email: dict["email"] as? String,
            phone: dict["phone"] as? String,
            avatarUrl: dict["avatar_url"] as? String ?? dict["avatar"] as? String,
            locale: dict["locale"] as? String,
            addresses: addrs,
            accounts: accs
        )
    }
}

struct ApiAddress {
    let label: String
    let street: String
    let number: String
    let complement: String?
    let neighborhood: String
    let city: String
    let state: String
    let country: String
    let zipCode: String
    let isPrimary: Bool

    var formatted: String { "\(street), \(number)" }
    var location: String { "\(city), \(state)" }

    static func from(_ dict: [String: Any]) -> ApiAddress? {
        ApiAddress(
            label: dict["label"] as? String ?? "",
            street: dict["street"] as? String ?? "",
            number: dict["number"] as? String ?? "",
            complement: dict["complement"] as? String,
            neighborhood: dict["neighborhood"] as? String ?? "",
            city: dict["city"] as? String ?? "",
            state: dict["state"] as? String ?? "",
            country: dict["country"] as? String ?? "",
            zipCode: dict["zip_code"] as? String ?? "",
            isPrimary: dict["is_primary"] as? Bool ?? false
        )
    }
}

struct ApiAccount {
    let provider: String
    let connectedAt: String

    static func from(_ dict: [String: Any]) -> ApiAccount? {
        guard let provider = dict["provider"] as? String else { return nil }
        return ApiAccount(
            provider: provider,
            connectedAt: dict["connected_at"] as? String ?? ""
        )
    }
}

// MARK: - Feed

struct FeedResponse {
    let posts: [ApiFeedPost]
    let total: Int
    let nextCursor: String?
    let hasMore: Bool

    static func from(_ dict: [String: Any]) throws -> FeedResponse {
        guard let postsArr = dict["posts"] as? [[String: Any]] else {
            throw ApiError.decode("Invalid feed response")
        }
        return FeedResponse(
            posts: postsArr.compactMap { ApiFeedPost.from($0) },
            total: dict["total"] as? Int ?? 0,
            nextCursor: dict["next_cursor"] as? String,
            hasMore: dict["has_more"] as? Bool ?? false
        )
    }
}

struct ApiFeedPost: Identifiable {
    let id: String
    let content: String
    let type: String
    let typeKey: String
    let typeColor: String?
    let categories: [String]
    let postedAt: String
    var signalsCount: Int
    let commentsCount: Int
    let distance: String?

    var timeAgo: String { formatTimeAgo(postedAt) }

    static func from(_ dict: [String: Any]) -> ApiFeedPost? {
        guard let id = dict["id"] as? String else { return nil }
        let cats = dict["categories"] as? [String] ?? []
        return ApiFeedPost(
            id: id,
            content: dict["content"] as? String ?? "",
            type: dict["type"] as? String ?? "",
            typeKey: dict["type_key"] as? String ?? "",
            typeColor: dict["type_color"] as? String,
            categories: cats,
            postedAt: dict["posted_at"] as? String ?? "",
            signalsCount: dict["signals_count"] as? Int ?? 0,
            commentsCount: dict["comments_count"] as? Int ?? 0,
            distance: dict["distance"] as? String
        )
    }
}

// MARK: - Comments

struct ApiComment: Identifiable {
    let id: String
    let postId: String
    let parentId: String?
    let authorName: String
    let text: String
    let upvotes: Int
    let downvotes: Int
    let createdAt: String
    let replies: [ApiComment]

    var timeAgo: String { formatTimeAgo(createdAt) }

    static func from(_ dict: [String: Any]) -> ApiComment? {
        guard let id = dict["id"] as? String else { return nil }
        let repliesArr = dict["replies"] as? [[String: Any]] ?? []
        return ApiComment(
            id: id,
            postId: dict["post_id"] as? String ?? "",
            parentId: dict["parent_id"] as? String,
            authorName: dict["author_name"] as? String ?? "",
            text: dict["text"] as? String ?? "",
            upvotes: dict["upvotes"] as? Int ?? 0,
            downvotes: dict["downvotes"] as? Int ?? 0,
            createdAt: dict["created_at"] as? String ?? "",
            replies: repliesArr.compactMap { from($0) }
        )
    }
}

// MARK: - Signals

struct SignalKey: Identifiable {
    let key: String
    let label: String
    let category: String
    let opposite: String?
    var id: String { key }

    static func from(_ dict: [String: Any]) -> SignalKey? {
        guard let key = dict["key"] as? String else { return nil }
        return SignalKey(
            key: key,
            label: dict["label"] as? String ?? key.replacingOccurrences(of: "_", with: " ").capitalized,
            category: dict["category"] as? String ?? "",
            opposite: dict["opposite"] as? String
        )
    }
}

struct SignalPair: Identifiable {
    let left: SignalKey
    let right: SignalKey
    var id: String { "\(left.key)-\(right.key)" }
}

struct SignalGroup: Identifiable {
    let category: String
    let label: String
    let pairs: [SignalPair]
    var id: String { category }

    static func groupFromKeys(_ keys: [SignalKey]) -> [SignalGroup] {
        let byCategory = Dictionary(grouping: keys, by: { $0.category })
        let categoryOrder = ["verification", "reaction"]
        let categoryLabels = ["verification": "Verificação", "reaction": "Reação"]

        return categoryOrder.compactMap { cat in
            guard let signals = byCategory[cat], !signals.isEmpty else { return nil }
            var pairs: [SignalPair] = []
            var used = Set<String>()
            for signal in signals {
                guard !used.contains(signal.key) else { continue }
                if let oppKey = signal.opposite,
                   let opp = signals.first(where: { $0.key == oppKey }) {
                    pairs.append(SignalPair(left: signal, right: opp))
                    used.insert(signal.key)
                    used.insert(opp.key)
                } else {
                    pairs.append(SignalPair(left: signal, right: signal))
                    used.insert(signal.key)
                }
            }
            return SignalGroup(category: cat, label: categoryLabels[cat] ?? cat.capitalized, pairs: pairs)
        }
    }
}

struct PostSignals {
    let signals: [String: Int]
    let mySignals: [String]

    static func from(_ dict: [String: Any]) -> PostSignals {
        let signalsDict = dict["signals"] as? [String: Int] ?? [:]
        let userArr = dict["my_signals"] as? [String] ?? dict["user_signals"] as? [String] ?? []
        return PostSignals(signals: signalsDict, mySignals: userArr)
    }
}

struct SyncSignalsResponse {
    let added: [String]
    let removed: [String]
    let signals: [String: Int]
    let mySignals: [String]

    static func from(_ dict: [String: Any]) -> SyncSignalsResponse {
        return SyncSignalsResponse(
            added: dict["added"] as? [String] ?? [],
            removed: dict["removed"] as? [String] ?? [],
            signals: dict["signals"] as? [String: Int] ?? [:],
            mySignals: dict["my_signals"] as? [String] ?? []
        )
    }
}

// MARK: - Reports

struct ReportReason: Identifiable {
    let key: String
    let label: String
    var id: String { key }

    static func from(_ dict: [String: Any]) -> ReportReason? {
        guard let key = dict["key"] as? String else { return nil }
        return ReportReason(
            key: key,
            label: dict["label"] as? String ?? key.replacingOccurrences(of: "_", with: " ").capitalized
        )
    }
}

// MARK: - Utils

private let isoFormatter: DateFormatter = {
    let f = DateFormatter()
    f.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
    f.timeZone = TimeZone(identifier: "UTC")
    f.locale = Locale(identifier: "en_US_POSIX")
    return f
}()

func formatCompactNumber(_ value: Int) -> String {
    switch value {
    case ..<1_000:
        return "\(value)"
    case ..<10_000:
        let k = Double(value) / 1_000
        return k.truncatingRemainder(dividingBy: 1) == 0
            ? "\(Int(k))k"
            : String(format: "%.1fk", k)
    case ..<1_000_000:
        return "\(value / 1_000)k"
    case ..<10_000_000:
        let m = Double(value) / 1_000_000
        return m.truncatingRemainder(dividingBy: 1) == 0
            ? "\(Int(m))M"
            : String(format: "%.1fM", m)
    default:
        return "\(value / 1_000_000)M"
    }
}

func toHashtag(_ text: String) -> String {
    let camel = text.split(separator: " ").map { word in
        word.prefix(1).uppercased() + word.dropFirst()
    }.joined()
    return "#\(camel)"
}

func formatTimeAgo(_ isoDate: String) -> String {
    guard !isoDate.isEmpty else { return "" }
    let clean = isoDate
        .components(separatedBy: ".").first?
        .components(separatedBy: "Z").first ?? isoDate
    guard let date = isoFormatter.date(from: clean) else { return "" }
    let diff = Int(Date().timeIntervalSince(date)) / 60
    switch diff {
    case ..<1: return "agora"
    case ..<60: return "\(diff)min"
    case ..<1440: return "\(diff / 60)h"
    case ..<43200: return "\(diff / 1440)d"
    default: return "\(diff / 43200)m"
    }
}

File: androidApp/src/main/kotlin/com/example/app/android/network/models/ApiModels.kt
package com.example.app.android.network.models

import org.json.JSONArray
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.TimeZone

// ── Auth ──

data class AuthResponse(
    val user: User,
    val accessToken: String,
    val refreshToken: String,
    val isNewUser: Boolean
) {
    companion object {
        fun fromJson(json: JSONObject) = AuthResponse(
            user = User.fromJson(json.getJSONObject("user")),
            accessToken = json.getString("access_token"),
            refreshToken = json.getString("refresh_token"),
            isNewUser = json.optBoolean("is_new_user", false)
        )
    }
}

data class TokenPair(
    val accessToken: String,
    val refreshToken: String
) {
    companion object {
        fun fromJson(json: JSONObject) = TokenPair(
            accessToken = json.getString("access_token"),
            refreshToken = json.getString("refresh_token")
        )
    }
}

data class User(
    val id: String,
    val name: String?,
    val displayName: String?,
    val email: String?,
    val phone: String?,
    val avatarUrl: String?,
    val addressCity: String?,
    val addressState: String?,
    val preferredLocale: String?,
    val addresses: List<UserAddress>,
    val accounts: List<UserAccount>
) {
    val initials: String get() {
        val parts = (name ?: displayName ?: "").split(" ")
        return if (parts.size >= 2) "${parts[0].take(1)}${parts[1].take(1)}".uppercase()
        else (name ?: displayName ?: "?").take(2).uppercase()
    }

    fun isProviderConnected(provider: String): Boolean =
        accounts.any { it.provider == provider }

    companion object {
        fun fromJson(json: JSONObject): User {
            val addrsArr = json.optJSONArray("addresses") ?: JSONArray()
            val addrs = (0 until addrsArr.length()).mapNotNull {
                UserAddress.fromJson(addrsArr.getJSONObject(it))
            }
            val accsArr = json.optJSONArray("accounts") ?: JSONArray()
            val accs = (0 until accsArr.length()).mapNotNull {
                UserAccount.fromJson(accsArr.getJSONObject(it))
            }
            return User(
                id = json.getString("id"),
                name = json.optString("name", null),
                displayName = json.optString("display_name", null) ?: json.optString("name", null),
                email = json.optString("email", null),
                phone = json.optString("phone", null),
                avatarUrl = json.optString("avatar_url", null),
                addressCity = json.optString("address_city", null),
                addressState = json.optString("address_state", null),
                preferredLocale = json.optString("preferred_locale", null),
                addresses = addrs,
                accounts = accs
            )
        }
    }
}

data class UserAddress(
    val label: String,
    val street: String,
    val number: String,
    val complement: String?,
    val neighborhood: String,
    val city: String,
    val state: String,
    val country: String,
    val zipCode: String,
    val isPrimary: Boolean
) {
    val formatted: String get() = "$street, $number"
    val location: String get() = "$city, $state"

    companion object {
        fun fromJson(json: JSONObject) = UserAddress(
            label = json.optString("label", ""),
            street = json.optString("street", ""),
            number = json.optString("number", ""),
            complement = json.optString("complement", null),
            neighborhood = json.optString("neighborhood", ""),
            city = json.optString("city", ""),
            state = json.optString("state", ""),
            country = json.optString("country", ""),
            zipCode = json.optString("zip_code", ""),
            isPrimary = json.optBoolean("is_primary", false)
        )
    }
}

data class UserAccount(
    val provider: String,
    val connectedAt: String
) {
    companion object {
        fun fromJson(json: JSONObject): UserAccount? {
            val provider = json.optString("provider", null) ?: return null
            return UserAccount(
                provider = provider,
                connectedAt = json.optString("connected_at", "")
            )
        }
    }
}

// ── Feed ──

data class FeedResponse(
    val posts: List<FeedPost>,
    val total: Int,
    val nextCursor: String?,
    val hasMore: Boolean
) {
    companion object {
        fun fromJson(json: JSONObject): FeedResponse {
            val postsArray = json.getJSONArray("posts")
            val posts = (0 until postsArray.length()).map {
                FeedPost.fromJson(postsArray.getJSONObject(it))
            }
            return FeedResponse(
                posts = posts,
                total = json.optInt("total", 0),
                nextCursor = json.optString("next_cursor", null),
                hasMore = json.optBoolean("has_more", false)
            )
        }
    }
}

data class FeedPost(
    val id: String,
    val content: String,
    val type: String,
    val typeKey: String,
    val typeColor: String?,
    val categories: List<String>,
    val postedAt: String,
    var signalsCount: Int,
    val commentsCount: Int,
    val distance: String?
) {
    val timeAgo: String get() = formatTimeAgo(postedAt)

    companion object {
        fun fromJson(json: JSONObject): FeedPost {
            val cats = json.optJSONArray("categories") ?: JSONArray()
            return FeedPost(
                id = json.getString("id"),
                content = json.optString("content", ""),
                type = json.optString("type", ""),
                typeKey = json.optString("type_key", ""),
                typeColor = json.optString("type_color", null),
                categories = (0 until cats.length()).map { cats.getString(it) },
                postedAt = json.optString("posted_at", ""),
                signalsCount = json.optInt("signals_count", 0),
                commentsCount = json.optInt("comments_count", 0),
                distance = json.optString("distance", null)
            )
        }
    }
}

// ── Post Creation ──

data class CreatedPost(
    val id: String,
    val originalText: String,
    val normalizedText: String,
    val type: String,
    val categories: List<String>,
    val location: PostLocation?
) {
    companion object {
        fun fromJson(json: JSONObject): CreatedPost {
            val cats = json.optJSONArray("categories")?.let { arr ->
                (0 until arr.length()).map {
                    val item = arr.get(it)
                    if (item is JSONObject) item.optString("label", "") else item.toString()
                }
            } ?: emptyList()
            return CreatedPost(
                id = json.getString("id"),
                originalText = json.optString("original_text", ""),
                normalizedText = json.optString("normalized_text", ""),
                type = json.optJSONObject("type")?.optString("label", "") ?: json.optString("type", ""),
                categories = cats,
                location = json.optJSONObject("location")?.let { PostLocation.fromJson(it) }
            )
        }
    }
}

data class PostLocation(
    val neighborhood: String?,
    val city: String?,
    val state: String?,
    val displayName: String?
) {
    companion object {
        fun fromJson(json: JSONObject) = PostLocation(
            neighborhood = json.optString("neighborhood", null),
            city = json.optString("city", null),
            state = json.optString("state", null),
            displayName = json.optString("display_name", null)
        )
    }
}

// ── Comments ──

data class Comment(
    val id: String,
    val postId: String,
    val parentId: String?,
    val authorName: String,
    val text: String,
    val upvotes: Int,
    val downvotes: Int,
    val createdAt: String,
    val replies: List<Comment>
) {
    val timeAgo: String get() = formatTimeAgo(createdAt)

    companion object {
        fun fromJson(json: JSONObject): Comment {
            val repliesArr = json.optJSONArray("replies") ?: JSONArray()
            return Comment(
                id = json.getString("id"),
                postId = json.optString("post_id", ""),
                parentId = json.optString("parent_id", null),
                authorName = json.optString("author_name", ""),
                text = json.optString("text", ""),
                upvotes = json.optInt("upvotes", 0),
                downvotes = json.optInt("downvotes", 0),
                createdAt = json.optString("created_at", ""),
                replies = (0 until repliesArr.length()).map {
                    fromJson(repliesArr.getJSONObject(it))
                }
            )
        }

        fun listFromJson(arr: JSONArray): List<Comment> =
            (0 until arr.length()).map { fromJson(arr.getJSONObject(it)) }
    }
}

// ── Signals ──

data class SignalKey(
    val key: String,
    val label: String,
    val category: String,
    val opposite: String?
) {
    companion object {
        fun fromJson(json: JSONObject) = SignalKey(
            key = json.getString("key"),
            label = json.optString("label", ""),
            category = json.optString("category", ""),
            opposite = json.optString("opposite", null)
        )
    }
}

data class SignalPair(
    val left: SignalKey,
    val right: SignalKey
)

data class SignalGroup(
    val category: String,
    val label: String,
    val pairs: List<SignalPair>
) {
    companion object {
        fun groupFromKeys(keys: List<SignalKey>): List<SignalGroup> {
            val byCategory = keys.groupBy { it.category }
            val categoryOrder = listOf("verification", "reaction")
            val categoryLabels = mapOf("verification" to "Verificação", "reaction" to "Reação")

            return categoryOrder.mapNotNull { cat ->
                val signals = byCategory[cat] ?: return@mapNotNull null
                if (signals.isEmpty()) return@mapNotNull null
                val pairs = mutableListOf<SignalPair>()
                val used = mutableSetOf<String>()
                for (signal in signals) {
                    if (signal.key in used) continue
                    val opp = signal.opposite?.let { oppKey -> signals.find { it.key == oppKey } }
                    if (opp != null) {
                        pairs.add(SignalPair(left = signal, right = opp))
                        used.add(signal.key)
                        used.add(opp.key)
                    } else {
                        pairs.add(SignalPair(left = signal, right = signal))
                        used.add(signal.key)
                    }
                }
                SignalGroup(category = cat, label = categoryLabels[cat] ?: cat.replaceFirstChar { it.uppercase() }, pairs = pairs)
            }
        }
    }
}

data class PostSignals(
    val signals: Map<String, Int>,
    val mySignals: List<String>
) {
    companion object {
        fun fromJson(json: JSONObject): PostSignals {
            val signalsObj = json.optJSONObject("signals") ?: JSONObject()
            val userArr = json.optJSONArray("my_signals") ?: json.optJSONArray("user_signals") ?: JSONArray()
            val map = mutableMapOf<String, Int>()
            signalsObj.keys().forEach { key -> map[key] = signalsObj.optInt(key, 0) }
            val userList = (0 until userArr.length()).map { userArr.getString(it) }
            return PostSignals(signals = map, mySignals = userList)
        }
    }
}

data class SyncSignalsResponse(
    val added: List<String>,
    val removed: List<String>,
    val signals: Map<String, Int>,
    val mySignals: List<String>
) {
    companion object {
        fun fromJson(json: JSONObject): SyncSignalsResponse {
            val addedArr = json.optJSONArray("added") ?: JSONArray()
            val removedArr = json.optJSONArray("removed") ?: JSONArray()
            val signalsObj = json.optJSONObject("signals") ?: JSONObject()
            val myArr = json.optJSONArray("my_signals") ?: JSONArray()
            val map = mutableMapOf<String, Int>()
            signalsObj.keys().forEach { key -> map[key] = signalsObj.optInt(key, 0) }
            return SyncSignalsResponse(
                added = (0 until addedArr.length()).map { addedArr.getString(it) },
                removed = (0 until removedArr.length()).map { removedArr.getString(it) },
                signals = map,
                mySignals = (0 until myArr.length()).map { myArr.getString(it) }
            )
        }
    }
}

// ── Reports ──

data class ReportReason(
    val key: String,
    val label: String
) {
    companion object {
        fun fromJson(json: JSONObject) = ReportReason(
            key = json.getString("key"),
            label = json.optString("label", "")
        )
    }
}

data class ReportResult(
    val reported: Boolean,
    val reportId: String?,
    val totalReports: Int,
    val postHidden: Boolean
) {
    companion object {
        fun fromJson(json: JSONObject) = ReportResult(
            reported = json.optBoolean("reported", false),
            reportId = json.optString("report_id", null),
            totalReports = json.optInt("total_reports", 0),
            postHidden = json.optBoolean("post_hidden", false)
        )
    }
}

// ── Utils ──

private val isoFormat = ThreadLocal.withInitial {
    SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }
}

fun toHashtag(text: String): String {
    val camel = text.split(" ").joinToString("") { word ->
        word.replaceFirstChar { it.uppercase() }
    }
    return "#$camel"
}

fun formatCompactNumber(value: Int): String = when {
    value < 1_000 -> "$value"
    value < 10_000 -> {
        val k = value / 100
        if (k % 10 == 0) "${k / 10}k" else "${k / 10}.${k % 10}k"
    }
    value < 1_000_000 -> "${value / 1_000}k"
    value < 10_000_000 -> {
        val m = value / 100_000
        if (m % 10 == 0) "${m / 10}M" else "${m / 10}.${m % 10}M"
    }
    else -> "${value / 1_000_000}M"
}

fun formatTimeAgo(isoDate: String): String {
    if (isoDate.isBlank()) return ""
    val date = try {
        isoFormat.get()!!.parse(isoDate.substringBefore('.').substringBefore('Z'))
    } catch (_: Exception) {
        return ""
    } ?: return ""
    val diffMs = System.currentTimeMillis() - date.time
    val minutes = diffMs / 60_000
    return when {
        minutes < 1 -> "agora"
        minutes < 60 -> "${minutes}min"
        minutes < 1440 -> "${minutes / 60}h"
        minutes < 43200 -> "${minutes / 1440}d"
        else -> "${minutes / 43200}m"
    }
}
